File: analyzer/src/main/kotlin/arch/analyzer/core/Ids.kt
package arch.analyzer.core

/**
 * Идентификаторы. Схема id операции — КОНТРАКТ (см. CONTRACTS.md и tools/gen-api.mjs):
 * на сгенерированные id ссылаются рукописные связи. Порт обязан быть побайтово
 * совместим с JS-реализацией; менять только синхронно с ней и с golden-тестами.
 */

/** {параметр} -> _p_, не-алфавитно-цифровое -> _, схлоп _+, срез хвостового _, обрезка 80. */
fun opId(method: String, path: String): String =
    (method.lowercase() + path
        .replace(Regex("\\{[^}]*}"), "_p_")
        .replace(Regex("[^a-zA-Z0-9]+"), "_")
        .replace(Regex("_+"), "_")
        .replace(Regex("_$"), ""))
        .take(80)

/** Ключ идентичности пути: имя параметра не различает эндпоинты. */
fun normPath(path: String): String = path.replace(Regex("\\{[^}]*}"), "{_p_}")

/** Слаг для id сторов/каналов: lowercase, [^a-z0-9]+ -> _, без краевых _. */
fun slug(s: String): String =
    s.lowercase()
        .replace(Regex("[^a-z0-9]+"), "_")
        .trim('_')

File: tools/ids.mjs
/**
 * Идентификаторы. Схема id операции — КОНТРАКТ: на сгенерированные id
 * ссылаются рукописные связи. Kotlin-порт (analyzer .../core/Ids.kt) обязан
 * давать те же id; менять только синхронно с ним и с golden-тестами.
 */

/** {параметр} -> _p_, не-алфавитно-цифровое -> _, схлоп _+, срез хвоста, обрезка 80. */
export const opId = (method, path) =>
  (
    method.toLowerCase() +
    path
      .replace(/\{[^}]*\}/g, '_p_')
      .replace(/[^a-zA-Z0-9]+/g, '_')
      .replace(/_+/g, '_')
      .replace(/_$/, '')
  ).slice(0, 80)

/** Слаг для id сторов/каналов: lowercase, [^a-z0-9]+ -> _, без краевых _. */
export const slug = (s) =>
  String(s)
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')

export const esc = (s) => String(s).replace(/'/g, "\\'")
